lib/global_uid/base.rb

```diff
@@ -29,169 +20,107 @@ def self.create_uid_tables(id_table_name, options={})
 
-      id_servers.map do |name, i|
-        info = {}
-        info[:cx]       = nil
-        info[:name]     = name
-        info[:retry_at] = nil
-        info[:rand]     = rand
-        info[:new?]     = true
-        info
-      end
-    end
-
-    def self.disconnect!
-      self.servers = nil
+      servers = id_servers.map { |name| GlobalUid::Server.new(name, increment_by: increment_by, connection_timeout: connection_timeout) }
+      # sorting here sets up each process to have affinity to a particular server.
+      servers.sort_by { |s| s.rand }
     end
 
     def self.setup_connections!
-      connection_timeout = self.global_uid_options[:connection_timeout]
-
-      if self.servers.nil?
-        self.servers = init_server_info
-        # sorting here sets up each process to have affinity to a particular server.
-        self.servers = self.servers.sort_by { |s| s[:rand] }
-      end
-
-      self.servers.each do |info|
-        next if info[:cx]
-
-        if info[:new?] || ( info[:retry_at] && Time.now > info[:retry_at] )
-          info[:new?] = false
-
-          connection = new_connection(info[:name], connection_timeout)
-          info[:cx]  = connection
-          info[:retry_at] = Time.now + self.global_uid_options[:connection_retry] if connection.nil?
+      self.servers ||= init_server_info
+
+      self.servers.each do |server|
+        next unless server.connection.nil?
+        if server.new? || server.retry_connection?
+          unless server.connect!
+            server.update_retry_at(GlobalUid.configuration.connection_retry)
+          end
         end
       end
-
-      self.servers
     end
 
-    def self.with_connections
+    def self.with_servers
       servers = setup_connections!
 
-      if !self.global_uid_options[:per_process_affinity]
+      if !GlobalUid.configuration.per_process_affinity?
         servers = servers.sort_by { rand } #yes, I know it's not true random.
       end
 
       raise NoServersAvailableException if servers.empty?
 
       errors = []
-      servers.each do |s|
+      servers.each do |server|
         begin
-          yield s[:cx] if s[:cx]
+          yield server unless server.connection.nil?
         rescue TimeoutException, Exception => e
-          notify e, "#{e.message}"
+          GlobalUid.configuration.notifier.notify(e, e.message)
           errors << e
-          s[:cx] = nil
-          s[:retry_at] = Time.now + 1.minute
+          server.connection = nil
+          server.update_retry_at(60)
         end
       end
 
       # in the case where all servers are gone, put everyone back in.
-      if servers.all? { |info| info[:cx].nil? }
-        servers.each do |info|
-          info[:retry_at] = Time.now - 5.minutes
-        end
+      if get_connections.empty?
+        servers.each { |server| server.update_retry_at(0) }
```

Comment:
I believe we can remove this line, it was added in the [initial commit](https://github.com/zendesk/global_uid/blob/2744dff891091053adfecb42718c6126c52acf78/lib/global_uid/base.rb#L145-L150), my understanding is that if a `NoServersAvailableException` exception is raised, the process/thread will die and it'll never retry the connection. It would have only worked before the gem was made thread safe in https://github.com/zendesk/global_uid/pull/21